### ISSUE TITLE
Is it ok to postpone writing test cases?

### ISSUE THREAD

**Issue by antonTan96** [nus-cs2103-AY2324S1/forum#281]
After experiencing v1.2, I realized that setting up stubs and writing utility classes for testing purposes takes up more time than writing code for the actual features. As a "feature freeze" is implemented in v1.4, we have to implement all the remaining features in v1.3. As I fear that there might not be enough time to extensively write test-cases for the features, is it ok to not write test cases and put them in in v1.4?

**Comment by damithc** [comment]
@antonTan96 This decision is up to you. We have no objections if you postpone some testing to v1.4 or give lower priority to testing in v1.3

**Comment by yiwen101** [comment]
Prof, may I clarify does that mean we could (if necessary) choose to merge PR not passing CI due low code coverage, and boost test coverage back in v1.4?

**Comment by damithc** [comment]
> Prof, may I clarify does that mean we could (if necessary) choose to merge PR not passing CI due low code coverage, and boost test coverage back in v1.4?

@yiwen101 Yes, definitely.

**Comment by yiwen101** [comment]
Thank you prof!

**Comment by s-peiran** [comment]
1. Just to clarify with testing during 1.4, that means we can add test methods and entire test.java classes in v1.4?
2. For resolving merge conflicts, is it okay to resolve in the Github web interface?

**Comment by damithc** [comment]
> * Just to clarify with testing during 1.4, that means we can add test methods and entire test.java classes in v1.4?

@s-peiran Yes.

> * For resolving merge conflicts, is it okay to resolve in the Github web interface?

Up to you. It is important to learn how to do it with Git (independent of GitHub, as projects you will do at your workplace might not use GitHub but very likely to use Git) but no harm learning both ways.

**Comment by eyelessrhyme7** [comment]
Feedback/Question/Open to Discussion: Given that the tP is also a breadth first approach, and we should aim to produce a working product at every version. Shouldn't it be a requirement to have the appropriate level of codecov throughout the versions, rather than delaying it to v1.4. I understand it may be because of the feature freeze process in v1.3 but it inadverdently still leads to piling up of the burden of testing and coverage to the last 10 days between v1.3 and v1.4.

**Comment by damithc** [comment]
> Feedback/Question/Open to Discussion: Given that the tP is also a breadth first approach, and we should aim to produce a working product at every version. Shouldn't it be a requirement to have the appropriate level of codecov throughout the versions, rather than delaying it to v1.4. I understand it may be because of the feature freeze process in v1.3 but it inadverdently still leads to piling up of the burden of testing and coverage to the last 10 days between v1.3 and v1.4.

@eyelessrhyme7 Yes, that is the recommended approach (see [here](https://nus-cs2103-ay2324s1.github.io/website/admin/tp-w10.html#:~:text=Recommend%3A%20Each%20PR%20should%20also%20update%20the%20relevant%20parts%20of%20documentation%20and%20tests)) but we don't enforce it, and it is probably not a good approach during early stages of a project when features are more volatile. It is still good to merge _some_ tests for critical parts of the code early rather than later. Also, as we enforce a feature freeze in v1.4, we have a buffer to pay off any accumulated testing 'debt'.
